File: src/main/java/com/dingdangmaoup/resin/pura/resin/configuration/JmxConfigurationStrategy.java
package com.dingdangmaoup.resin.pura.resin.configuration;

import com.dingdangmaoup.resin.pura.ResinModelBase;
import com.dingdangmaoup.resin.pura.resin.WebApp;
import com.intellij.javaee.appServers.deployment.DeploymentStatus;
import com.intellij.openapi.util.Ref;
import org.jetbrains.annotations.NotNull;

public interface JmxConfigurationStrategy {

  boolean deployWithJmx(ResinModelBase resinModel, WebApp webApp);

  boolean undeployWithJmx(ResinModelBase resinModel, WebApp webApp);

  @NotNull
  DeploymentStatus getDeployStateWithJmx(ResinModelBase resinModel, WebApp webApp, Ref<Boolean> isFinal);
}

File: src/main/kotlin/com/dingdangmaoup/resin/pura/resin/configuration/JmxConfigurationStrategy.kt
package com.dingdangmaoup.resin.pura.resin.configuration

import com.dingdangmaoup.resin.pura.ResinModelBase
import com.dingdangmaoup.resin.pura.resin.WebApp
import com.dingdangmaoup.resin.pura.resin.DeploymentObservation
import com.intellij.javaee.appServers.deployment.DeploymentStatus
import com.intellij.openapi.util.Ref

interface JmxConfigurationStrategy {
    fun deployWithJmx(resinModel: ResinModelBase<*>, webApp: WebApp): Boolean

    fun undeployWithJmx(resinModel: ResinModelBase<*>, webApp: WebApp): Boolean

    fun getDeployStateWithJmx(resinModel: ResinModelBase<*>, webApp: WebApp, isFinal: Ref<Boolean>): DeploymentStatus

    fun observeConfiguration(resinModel: ResinModelBase<*>, webApp: WebApp, removing: Boolean): DeploymentObservation

    fun observeArchive(resinModel: ResinModelBase<*>, webApp: WebApp, removing: Boolean): DeploymentObservation
}
